packages/flutter_tools/lib/src/crash_reporting.dart

```diff
@@ -0,0 +1,383 @@
+/// [files].
+///
+/// This request automatically sets the Content-Type header to
+/// `multipart/form-data`. This value will override any value set by the user.
+///
+///     var uri = Uri.parse("http://pub.dartlang.org/packages/create");
+///     var request = new http.MultipartRequest("POST", url);
+///     request.fields['user'] = 'nweiz@google.com';
+///     request.files.add(new http.MultipartFile.fromFile(
+///         'package',
+///         new File('build/package.tar.gz'),
+///         contentType: new MediaType('application', 'x-tar'));
+///     request.send().then((response) {
+///       if (response.statusCode == 200) print("Uploaded!");
+///     });
+class _MultipartRequest extends http.BaseRequest {
+  /// The total length of the multipart boundaries used when building the
+  /// request body. According to http://tools.ietf.org/html/rfc1341.html, this
+  /// can't be longer than 70.
+  static const int _BOUNDARY_LENGTH = 70;
+
+  static final Random _random = new Random();
+
+  /// The form fields to send for this request.
+  final Map<String, String> fields;
+
+  /// The private version of [files].
+  final List<http.MultipartFile> _files;
+
+  /// Creates a new [MultipartRequest].
+  _MultipartRequest(String method, Uri url)
+      : fields = <String, String>{},
+        _files = <http.MultipartFile>[],
+        super(method, url);
+
+  /// The list of files to upload for this request.
+  List<http.MultipartFile> get files => _files;
+
+  /// The total length of the request body, in bytes. This is calculated from
+  /// [fields] and [files] and cannot be set manually.
+  @override
+  int get contentLength {
+    int length = 0;
+
+    fields.forEach((String name, String value) {
+      length += "--".length + _BOUNDARY_LENGTH + "\r\n".length +
+          UTF8.encode(_headerForField(name, value)).length +
+          UTF8.encode(value).length + "\r\n".length;
+    });
+
+    for (http.MultipartFile file in _files) {
+      length += "--".length + _BOUNDARY_LENGTH + "\r\n".length +
+          UTF8.encode(_headerForFile(file)).length +
+          file.length + "\r\n".length;
+    }
+
+    return length + "--".length + _BOUNDARY_LENGTH + "--\r\n".length;
+  }
+
+  @override
+  set contentLength(int value) {
+    throw new UnsupportedError("Cannot set the contentLength property of "
+        "multipart requests.");
+  }
+
+  /// Freezes all mutable fields and returns a single-subscription [ByteStream]
+  /// that will emit the request body.
+  @override
+  http.ByteStream finalize() {
+    String boundary = _boundaryString();
+    headers['content-type'] = 'multipart/form-data; boundary=$boundary';
+    super.finalize();
+
+    StreamController<List<int>> controller = new StreamController<List<int>>(sync: true);
+
+    void writeAscii(String string) {
+      controller.add(UTF8.encode(string));
+    }
+
+    void writeUtf8(String string) {
+      controller.add(UTF8.encode(string));
+    }
+    void writeLine() {
```

Comment:
While I cleaned up the code to satisfy our analyzer lints, I intentionally kept changes to a minimum so that it's still recognizable when compared to the original `package:http` implementation.